File: 1_Arrays/Day_008_StockButAndSell_MaxOneTransactionAllowed/Solution.py
class Solution:
    '''
        Approach: one Pass
        1. Initialize lMin to the first element of prices and profit to 0.
        2. Iterate through the prices starting from the second element.
        3. For each price, update lMin to be the minimum of lMin and the current price.
        4. Calculate the potential profit by subtracting lMin from the current price.
        5. Update profit to be the maximum of profit and the potential profit.
        6. Return profit.
        Time Complexity: O(n), where n is the number of prices.
        Space Complexity: O(1), as we are using only a constant amount of space.
    '''
    # Function to find the maximum profit from stock prices with one transaction allowed.
    def maximumProfit(self, prices):
        n = len(prices)
        # Initialize lMin to the first element of prices and profit to 0.
        lMin = prices[0]
        profit = 0
        # Iterate through the prices starting from the second element.
        # For each price, update lMin to be the minimum of lMin and the current price.
        for i in range(1,n):
            # Calculate the potential profit by subtracting lMin from the current price.
            lMin = min(lMin, prices[i])
            # Update profit to be the maximum of profit and the potential profit.
            profit = max(profit, prices[i] - lMin)
        # Return profit.
        return profit

File: 1_Arrays/Day_008_StockButAndSell_MaxOneTransactionAllowed/Solution.java
class Solution {

    /**
     * Approach: One Pass
     * Function to find the maximum profit from stock prices
     * The function iterates through the array of prices and keeps track of the minimum price seen so far.
     * It calculates the profit by subtracting the minimum price from the current price and updates the maximum profit if necessary.
     * Finally, it returns the maximum profit.
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     * @param prices
     * @return
     */
    // Function to find the maximum profit from stock prices
    public int maximumProfit(int prices[]) {
        // Initialize the minimum price to the first element of the array
        int lMin = prices[0];
        // Initialize the maximum profit to 0
        int profit = 0;
        // Iterate through the array starting from the second element
        // Update the minimum price and maximum profit as necessary
        for(int i=1; i<prices.length; i++){
            // Update the minimum price if the current price is lower
            lMin = Math.min(lMin, prices[i]);
            // Calculate the profit by subtracting the minimum price from the current price
            // Update the maximum profit if the current profit is higher
            profit = Math.max(profit, (prices[i] - lMin));
        }
        // Return the maximum profit
        // The maximum profit is the maximum difference between the current price and the minimum price seen so far
        return profit;
    }
}